Faster boot times and Reproducible builds

newtonick opened the pull request:
This pull request has 2 intermingled functional changes. Reducing the time from power on to splash screen and a method for producing deterministic reproducible builds. The main reason these changes are intermingled is because both required the reduction of build dependencies. Specific reproducible steps could be pulled out into its own, but this felt like unnecessary and time consuming work.

Faster boot times don’t need much explanation. I measure the pi0 board booting in 15 seconds instead of 45 in release 0.6.0. The image size is 27262976 bytes total (compared to the previous release at 35652096 bytes, an 8MB of savings).

Having a SeedSigner OS reproducible build allows creation of the image to be compiled and assembled deterministically. This makes it possible for multiple people to compile the same source code into identical binaries. This also means sole trust in SeedSigner the person (or any other SeedSigner contributors) isn’t needed for the image that most people run on thier SeedSigner. This trust can be distributed through others attesting they to can produce the same image byte for byte. Also for those who compile the image themselves, they have a way to verify the image produced is identical to the one released.

Here is a breakdown of configuration and script changes made to accomplish these 2 larger functional goals.

* Update buildroot sub module to more modern version (2023.02). I didn’t want to do a deep dependency dive on an older version and then later upgrade. 
* Use buildroot fork controlled by seedsigner account. This patched fork forces use of python 3.10.10 and does the python build in a reproducible way.
* Remove python-qrcode external package because it’s now a buildroot standard package.
* Update build.sh script to replace --keep-alive and --skip-build with --no-op option. The `--no-op` option is the new default that performs no action but keeps the docker container running indefinitely.
* Update build.sh script to allow checkout of a specific app commit and defaults to the dev branch. 
* Update docker-compose to pull specific apt packages
* Update README
* Other minor cleanup items

This PR replaces #45 and #34.

Written instructions for how to create reproducible builds:
https://gist.github.com/newtonick/2df134a84fd04398bd925a000e979112

Known issues with this PR:
* USB Keyboard does not work in Pi4 dev build
* Mdev does not work in Pi4 and Pi4 dev builds

newtonick commented:
My reproducible build testing results. All builds done with amd64 architecture. I will update this table as I complete builds.

| Build Script Options            | Image File Name                   | sha256  Output                                              |
| ------------------------------- | --------------------------------- | ---------------------------------------------------------------- |
|`--pi0 --app-commit-id=9c36f5c`  |`seedsigner_os.9c36f5c.pi0.img`    | `29e41c ... 28b036` |
|`--pi2 --app-commit-id=9c36f5c`  |`seedsigner_os.9c36f5c.pi2.img`    | `2f470d ... 85e6a9` |

kdmukai commented:
Success on M1 Pro!!!

<img width="1250" alt="Screenshot 2023-07-18 at 12 15 50 AM" src="https://github.com/SeedSigner/seedsigner-os/assets/934746/5bac08d5-88a0-45d1-b7f4-18f11e66f19b">

2hrs 38min

Note on macOS it's `shasum -a 256 [file]`

jdlcdl commented:
Your gist is such detailed work that I think I learned more about building seedsigner-os in this one document than I have in the past.  I love the alternate methods and explanations.  I'll note that I recall a discussion like: "Maybe a pre-release and docs for public scrutiny aren't necessary upfront?" and "Maybe not a show-stopper if pi2 and pi4 boards are not supported right away?"... and then you went on to slay them both... all reproducible, all exactly the same file sizes, Wow!


As of your gist at "Revisions=24"...

I've attempted a proof-read from top to bottom: minor typos:
* "others attesting they to can produce" becomes "others attesting they too can produce"
* "You should now have an shell session open" becomes "You should now have a shell session open"
* "You'll need to clone to app repo first" becomes "You'll need to clone the app repo first"
* "This will produce a produce a list" becomes "This will produce a list"
* header "install Docker and my recommend configurations" becomes "install Docker and my recommended configurations"

I've followed the instructions, cutting and pasting from your gist, w/o major problems:
* pi0, pi02w and pi2 built w/ "You want more?" instructions (last 2 w/ modified build.sh commandlines).
* pi4 built from same container, except w/ "Still not satisfied?" instructions and modified `make` commandline.
There was visible error-output in the `find...-delete` command: 
`find: cannot delete ‘/opt/rootfs-overlay/opt’: Directory not empty`
... but it did do it's job -- only `src/` remained.
Because I had just built for pi2 and output/ was there, I had to `rm -rf /output` inside the container, then run the two `make` commands again, modifying the first for pi4.
Afterwards, I did `cp /output/images/seedsigner-os.img /images/` and renamed it to be more appropriate for my pi4 build.
* the "Bonus Round" commands all worked. For a quick comparison, the sha256 hash of my `rootfs.md5.chk` file is 
`e9745263f0d149a50b109c29f230092cdefe74a9e15ef5c33dca663c6fdb5edb`.
...but I did get a single line of diff output after comparing. Yours has:
`d41d8cd98f00b204e9800998ecf8427e  ./dev/console` while mine does not; all others are same in rootfs for the pi0 image. (might your's have been from a previous build?)
* each build ran in just under 40m on Ubuntu/x86_64 system w/ 4-core-Xeon and 16GB RAM -- via user-space docker.
* I did NOT try the `docker compose up` one-liner (because I don't have docker-compose).

Reproduced images for all supported boards appear to match your sha256 hashes; as for this attestation:
```
$ sha256sum ./images/*
f99d1b27b77264319fbc747c9b3e47c617445002826b248ae2cb8ce9b9be20bc  ./images/seedsigner_os.9c36f5c.pi02w.img
29e41cccf0ca93705901bacd5d605da884f9b8315cfd978910698c2f8e28b036  ./images/seedsigner_os.9c36f5c.pi0.img
2f470dfca0491d97696ccdd9c45104cb1bab0af52a8405e85ef7fbc7cf85e6a9  ./images/seedsigner_os.9c36f5c.pi2.img
d295148e754f48504f559daf50ac92ed8ca2b94b2b22bfbd214cd49146ee4c5c  ./images/seedsigner_os.9c36f5c.pi4.img
```

The images for the pi0, pi2, and pi4 all boot fast and succeed the "I/O test" (w/ camera_rotation=0 as expected).
* pi0 splash at ~15s, main menu at ~26s
* pi2 splash at ~11s, main menu at ~20s
* pi4 splash at ~16s, main menu at ~23s
* I did NOT test the pi02w image.

kdmukai commented:
Notes for running in a LunaNode VPS:

Why do the reproducible build in a Virtual Machine(VM)? 
The build should be straightforward on a Linux OS but is potentially hopeless on Windows. Windows users are much better off avoiding the Windows complications altogether and building instead on a brand new Ubuntu VM that starts pristine and will be discarded as soon as the build is done.

Why LunaNode?
* It's a very easy UI to spin up a new virtual machine (VM).
* aws is overkill and too complicated for this kind of simple task.
* They have a free trial for new users (almost definitely sufficient for our build).
* Can initially fund an account with as little as $5 (that's way more than enough to do the build)
* Accepts Lightning!


## Create LunaNode instance:
* Create an account w/email addr
* Use their free trial or fund account with $5 in Lightning via their btcpay server
* Click "Create VM" in the left nav menu
* Select region: Toronto (only option available)
* Hostname: anything you want (e.g. "seedsigner_builder")
* Plan: Select the "Compute Optimized" instance category and choose at least "c.2".
* Image: Ubuntu-22.04

Once it's created:
* click into the new instance details
* go to the "CPU" tab
* Enable "Pay for CPU utilization above baseline performance?"
* go to the "General" tab
* note its External IP
* note the password in "Initial login details".

## ssh into VM
From a local terminal: `ssh ubuntu@[server's public ip]` and use the password from above when prompted.

## Install Docker
Once inside the new VM, install Docker via the convenience script: https://docs.docker.com/engine/install/ubuntu/#install-using-the-convenience-script
(if the script fails with "Unable to acquire the dpkg frontend lock", just wait a minute or two; the new VM is still running its initial setup steps)

Or if you want more control, you can do a manual install:
https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository

## Configure Docker permissions
The convenience script recommends a follow-up step to take care of this part:
```bash
sudo apt-get install -y uidmap
dockerd-rootless-setuptool.sh install
```

Or you can do it manually:
Steps 1-4: https://docs.docker.com/engine/install/linux-postinstall/

## Confirm your Docker setup
`docker run hello-world`

## Build SeedSigner-OS image
(follow @newtonick's notes from there; shouldn't need any modifications now that we're in a totally standard Ubuntu VM)

Build time on a c.2 image: (still working)

## Copying the resulting .img off the VM
??

Shouldn't be necessary if the hashes match; you can be confident that downloading the release image from the SeedSigner github repo is identical to building it yourself.

## Delete the LunaNode VM instance
In the LunaNode "Virtual Machines" page, click on your instance, and then click the red "Delete" button at the top right and follow the instructions.

jdlcdl commented on README.md:
```diff
@@ -101,24 +102,37 @@ Here is a table Raspberry Pi boards to image filenames/configs
 ### Development cycle using docker
 
-Each time the `docker-compose up` command runs a full build from scratch is performed. You can optionally run `docker-compose up -d` in detached mode by adding the `-d` flag. This will run the container in the background. To have faster development cycles you'll likely want to avoid building the OS from scratch each time. You can avoid recreating the docker image/container a few different ways. O ...
+Each time the `docker compose up` command runs a full build from scratch is performed. You can optionally run `docker compose up -d` in detached mode by adding the `-d` flag. This will run the container in the background. To have faster development cycles you'll likely want to avoid building the OS from scratch each time. You can avoid recreating the docker image/container a few different ways. O ...
```
3 sentences from end of paragraph: "docker-compse up" becomes "docker compose up"

jdlcdl commented on README.md:
```diff
@@ -101,24 +102,37 @@ Here is a table Raspberry Pi boards to image filenames/configs
+Each time the `docker compose up` command runs a full build from scratch is performed. You can optionally run `docker compose up -d` in detached mode by adding the `-d` flag. This will run the container in the background. To have faster development cycles you'll likely want to avoid building the OS from scratch each time. You can avoid recreating the docker image/container a few different ways. O ...
 
-Using docker-compose will build the image and launch the container
+Using docker compose will start the container (create new container if on does not already exist) without building an image
```
"create new container if on does not" becomes "creating a new container if one does not"

gpatkinson commented:
the hash matches on my Ubuntu LTS 20.04 VM. fantastic work, @newtonick !

![image](https://github.com/SeedSigner/seedsigner-os/assets/8332802/257b309a-0250-4c0b-8726-d51d230a41ca)

newtonick commented:
> * "others attesting they to can produce" becomes "others attesting they too can produce"
> * "You should now have an shell session open" becomes "You should now have a shell session open"
> * "You'll need to clone to app repo first" becomes "You'll need to clone the app repo first"
> * "This will produce a produce a list" becomes "This will produce a list"
> * header "install Docker and my recommend configurations" becomes "install Docker and my recommended configurations"

Thank you for these, I have applied these updates to the gist.



> Your gist is such detailed work that I think I learned more about building seedsigner-os in this one document than I have in the past. I love the alternate methods and explanations. I'll note that I recall a discussion like: "Maybe a pre-release and docs for public scrutiny aren't necessary upfront?" and "Maybe not a show-stopper if pi2 and pi4 boards are not supported right away?"... and then you went on to slay them both... all reproducible, all exactly the same file sizes, Wow!
> 
> As of your gist at "Revisions=24"...
> 
> I've attempted a proof-read from top to bottom: minor typos:
> 
> * "others attesting they to can produce" becomes "others attesting they too can produce"
> * "You should now have an shell session open" becomes "You should now have a shell session open"
> * "You'll need to clone to app repo first" becomes "You'll need to clone the app repo first"
> * "This will produce a produce a list" becomes "This will produce a list"
> * header "install Docker and my recommend configurations" becomes "install Docker and my recommended configurations"
> 
> I've followed the instructions, cutting and pasting from your gist, w/o major problems:
> 
> * pi0, pi02w and pi2 built w/ "You want more?" instructions (last 2 w/ modified build.sh commandlines).
> * pi4 built from same container, except w/ "Still not satisfied?" instructions and modified `make` commandline.
>   There was visible error-output in the `find...-delete` command:
>   `find: cannot delete ‘/opt/rootfs-overlay/opt’: Directory not empty`
>   ... but it did do it's job -- only `src/` remained.
>   Because I had just built for pi2 and output/ was there, I had to `rm -rf /output` inside the container, then run the two `make` commands again, modifying the first for pi4.
>   Afterwards, I did `cp /output/images/seedsigner-os.img /images/` and renamed it to be more appropriate for my pi4 build.
> * the "Bonus Round" commands all worked. For a quick comparison, the sha256 hash of my `rootfs.md5.chk` file is
>   `e9745263f0d149a50b109c29f230092cdefe74a9e15ef5c33dca663c6fdb5edb`.
>   ...but I did get a single line of diff output after comparing. Yours has:
>   `d41d8cd98f00b204e9800998ecf8427e  ./dev/console` while mine does not; all others are same in rootfs for the pi0 image. (might your's have been from a previous build?)
> * each build ran in just under 40m on Ubuntu/x86_64 system w/ 4-core-Xeon and 16GB RAM -- via user-space docker.
> * I did NOT try the `docker compose up` one-liner (because I don't have docker-compose).
> 
> Reproduced images for all supported boards appear to match your sha256 hashes; as for this attestation:
> 
> ```
> $ sha256sum ./images/*
> f99d1b27b77264319fbc747c9b3e47c617445002826b248ae2cb8ce9b9be20bc  ./images/seedsigner_os.9c36f5c.pi02w.img
> 29e41cccf0ca93705901bacd5d605da884f9b8315cfd978910698c2f8e28b036  ./images/seedsigner_os.9c36f5c.pi0.img
> 2f470dfca0491d97696ccdd9c45104cb1bab0af52a8405e85ef7fbc7cf85e6a9  ./images/seedsigner_os.9c36f5c.pi2.img
> d295148e754f48504f559daf50ac92ed8ca2b94b2b22bfbd214cd49146ee4c5c  ./images/seedsigner_os.9c36f5c.pi4.img
> ```
> 
> The images for the pi0, pi2, and pi4 all boot fast and succeed the "I/O test" (w/ camera_rotation=0 as expected).
> 
> * pi0 splash at ~15s, main menu at ~26s
> * pi2 splash at ~11s, main menu at ~20s
> * pi4 splash at ~16s, main menu at ~23s
> * I did NOT test the pi02w image.

Great comments and feedback. Thank you for the thorough testing!

jahangir13 commented:
Worked in my seedsigner Debian 11 lxd container for the pi0 image:
Took about 1 1/2 hours with 2 CPUs assigned (nearly all the time at 99%). Disk space needed for Debian and image build: about 14GB
![grafik](https://github.com/SeedSigner/seedsigner-os/assets/7389834/30695813-9840-4aad-8818-bda0a121ae2b)

jahangir13 commented:
Worked also in my Windows 10 (WSL2/Ubuntu Terminal) for the pi0 image with the one-liner command (now less than 40 minutes with 6 CPU cores):

![grafik](https://github.com/SeedSigner/seedsigner-os/assets/7389834/d7eacfdc-e6ad-46b5-8cc7-b75ef406c54f)

kdmukai commented:
@jahangir13 maybe my mistake was trying to have Docker Desktop in Windows run the Docker service for WSL2.

I'll see if I can go back and stay purely in Ubuntu/WSL2.

jahangir13 commented:
> @jahangir13 maybe my mistake was trying to have Docker Desktop in Windows run the Docker service for WSL2.
> 
> I'll see if I can go back and stay purely in Ubuntu/WSL2.

Yes, exactly. I just opened the Ubuntu Terminal (I did not start Docker Desktop at all). Then installed docker stuff in the Terminal (so in Ubuntu). I now also give up with Powershell for now,

kdmukai commented:
I also discovered that when Docker Desktop had the WSL2 integration enabled, it was slamming my loud HDD zzt-zzt-zzt-zzt nonstop. For hours. Likely forever.

As soon as I stopped Docker Desktop, all those read/writes stopped. When I disabled the WSL2 integration and Docker Desktop restarted, it left my HDD in peace.

kdmukai commented:
Another confirmation, via twitter (yes, weight it accordingly)

https://twitter.com/4moonsettler/status/1682057433298894852

kdmukai commented:
Big f'n ACK.

I'm satisfied that it works on enough platforms. Documentation and step-by-step guidance (esp for Windows) still needed, but recommend that be done in follow up PRs (and can be submitted by more contributors to distribute the workload).

jdlcdl commented:
I've finally succeeded with the one-liner method, for pi2-dev.  sha256sum looks like:
`b7f738035f4794b0f8b50afd07a25d980345868690864830ea3867ab05dfca00  ../seedsigner-os/images/seedsigner_os.9c36f5c.pi2-dev.img`

ACK

newtonick commented:
On the VPS topic, just another example of building on VPS. Here is a Linode VPS script I have used to build a test SeedSigner OS image. It's 100% command line through Linodes command line client. I'm sharing it because it's an example of building the SeedSigner OS image without docker. 

### Create Linode VPS

```bash
export LINODE_TMP_ROOT_PASSWD=$(head /dev/urandom | LC_ALL=C tr -dc A-Za-z0-9 | head -c 20)
echo "Linode Root Password: ${LINODE_TMP_ROOT_PASSWD}"
export LINODE_TMP_SSH_AUTH_KEY=$(cat ~/.ssh/id_ed25519.pub)

linode-cli linodes create \
 --type g6-dedicated-8 \
 --region us-east \
 --image linode/debian11 \
 --root_pass ${LINODE_TMP_ROOT_PASSWD} \
 --authorized_keys "${LINODE_TMP_SSH_AUTH_KEY}" \
 --label seedsigner-os

export LINODE_SEEDSIGNER_OS_IP=$(linode-cli linodes ls --label seedsigner-os --json | jq -r '.[0].ipv4[0]')
```

### Connect via SSH to VPS once server is provisioned and starts sshd service

```bash
ssh -o "UserKnownHostsFile=/dev/null" -o "StrictHostKeyChecking=no" root@${LINODE_SEEDSIGNER_OS_IP}
```

### Install Docker, Docker Compose and then Build the SeedSigner OS Image on VPS

```bash
apt update -qq
apt install  -qq -y ca-certificates curl gnupg

# Add Docker’s official GPG key:
install -m 0755 -d /etc/apt/keyrings
curl -fsSL https://download.docker.com/linux/debian/gpg | gpg --dearmor -o /etc/apt/keyrings/docker.gpg
chmod a+r /etc/apt/keyrings/docker.gpg

# Use the following command to set up the repository:
echo \
  "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian \
  "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable" | \
  tee /etc/apt/sources.list.d/docker.list > /dev/null

# Install Docker Engine and Docker Compose.
apt update -qq
sudo apt-get install -y docker-ce docker-ce-cli docker-compose-plugin

# Build SeedSigner OS Image
git clone --recursive https://github.com/SeedSigner/seedsigner-os.git
cd seedsigner-os
git fetch origin pull/51/head:pr_51
git checkout pr_51
git submodule update
DOCKER_DEFAULT_PLATFORM="linux/amd64" SS_ARGS="--pi0 --app-commit-id=9c36f5c" docker compose up --force-recreate --build
uptime
exit
```

### Delete Linode VPS when done

```bash
export LINODE_SEEDSIGNER_OS_ID=$(linode-cli linodes ls --label seedsigner-os --json | jq -r '.[0].id')
linode-cli linodes delete ${LINODE_SEEDSIGNER_OS_ID}
```

### Output, last few lines for reference

42 minutes of uptime to build in this one example

```bash
seedsigner-os-build-images-1  | /opt/buildroot
seedsigner-os-build-images-1  | 29e41cccf0ca93705901bacd5d605da884f9b8315cfd978910698c2f8e28b036  /opt/../images/seedsigner_os.9c36f5c.pi0.img
seedsigner-os-build-images-1 exited with code 0
root@localhost:~/seedsigner-os# uptime
 21:06:08 up 42 min,  1 users,  load average: 3.77, 4.89, 4.32
root@localhost:~/seedsigner-os# exit
logout
Connection to 50.116.57.212 closed.
user@machine ~ % export LINODE_SEEDSIGNER_OS_ID=$(linode-cli linodes ls --label seedsigner-os --json | jq -r '.[0].id')
linode-cli linodes delete ${LINODE_SEEDSIGNER_OS_ID}
```

hugoender commented:
This is amazing! If only all open source projects had this type of thorough documentation for their repro builds. Amazing work @newtonick

Marc-Gee commented:
**Windows Success** using docker desktop !! (via the single line command )
The pi02w hash matches ! 
f99d1b27b77264319fbc747c9b3e47c617445002826b248ae2cb8ce9b9be20bc
<img width="1009" alt="image" src="https://github.com/SeedSigner/seedsigner-os/assets/91296549/892d518c-406e-4138-a28b-de1e2cfeafb3">

hugoender commented:
As I was running the following command `./build.sh --pi0 --app-commit-id=9c36f5c` I noticed in the very numerous text that was scrolling through the terminal that there was a connection to `http://mirrors.tripadvisor.com/gnu/mpc/mpc-1.2.1.tar.gz` in there (see last lines of screenshot below). Why is there a connection to tripadvisor.com?!

![Screenshot from 2023-07-21 21-43-24](https://github.com/SeedSigner/seedsigner-os/assets/6217175/65cc68ee-c66c-432d-9b53-c6303c25b593)

jahangir13 commented:
> `http://mirrors.tripadvisor.com/gnu/mpc/mpc-1.2.1.tar.gz`

They provide a GNU mirror server (as you also could do). 
These can be seen in this list: https://www.gnu.org/prep/ftp.en.html

And depending on your location and network state it selects one from the list: 
![grafik](https://github.com/SeedSigner/seedsigner-os/assets/7389834/895f89ac-17e9-4c68-8414-cb36a8dafc29)

hugoender commented:
> > `http://mirrors.tripadvisor.com/gnu/mpc/mpc-1.2.1.tar.gz`
> 
> They provide a GNU mirror server (as you also could do). These can be seen in this list: https://www.gnu.org/prep/ftp.en.html
> 
> And depending on your location and network state it selects one from the list: ![grafik](https://user-images.githubusercontent.com/7389834/255332930-895f89ac-17e9-4c68-8414-cb36a8dafc29.png)
> 
> At least this is my understanding ;)

That is VERY interesting. Never thought a company like TripAdvisor would provide a GNU mirror. Is there any benefit to them to do that or would it just be something they do because they believe in the project?

hugoender commented:
Feedback on this PR. The instructions are great but there is one small detail missing for those of us that are not very familiar with Docker: how do you pull or transfer the image that's created after `./build.sh --pi0 --app-commit-id=9c36f5c`  from the Docker container to the host? Is this even necessary or can you flash from within Docker container?

jdlcdl commented on opt/build.sh:
```diff
@@ -52,10 +63,13 @@ download_app_repo() {
   rm -rf ${rootfs_overlay}/opt/setup.py
+  rm -rf ${rootfs_overlay}/opt/tests
+  rm -rf ${rootfs_overlay}/opt/tools
+  rm -rf ${rootfs_overlay}/opt/pytest.ini
```
can also add requirements-raspi.txt to list of files to delete... since flow-tests merged recently.

newtonick replied:
I'll leave this for a followup PR.
